### PR TITLE
Log1p for complex in CPU 

### PR DESCRIPTION
**Pull request by mfkasim1**
Another PR for https://github.com/pytorch/pytorch/issues/89205: making torch.log1p accepts complex numbers in CPU.
I haven't done the GPU version because I'm not sure which file(s) to change.

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: test/test_unary_ufuncs.py [hunk 1]
```diff
@@ -1094,6 +1094,43 @@ def test_mish(self, device, dtype):
...
+    @dtypes(torch.complex64, torch.complex128)
+    @onlyCPU
+    def test_log1p_complex(self, device, dtype):
+        inouts = [
+            (0.2 + 0.3j, 0.21263386770217202 + 0.24497866312686414j),
+            (1e-19 + 1e-18j, 1e-19 + 1e-18j),
+            (1e-18 + 0.1j, 0.00497517 + 0.0996687j),
+            (0.1 + 1e-18j, 0.0953102 + 9.090909090909090909e-19j),
+            (0.5 + 0j, 0.40546510810816 + 0j),
+            (0.0 + 0.5j, 0.111571776 + 0.463647609j),
```
Comment (jgong5) [inline test/test_unary_ufuncs.py:1106]: Consider using numpy as a ref check instead hard-coding the ref result?

Comment (mfkasim1) [reply test/test_unary_ufuncs.py:1106]: numpy's `log1p` suffers from loss of precision with relatively small complex number (see [here](https://github.com/numpy/numpy/pull/22611)). That's why I'm using the hard-coded results, calculated from arbitrary-precision math and wolfram alpha.

Comment (mruberry) [reply test/test_unary_ufuncs.py:1106]: Writing a test like this is so unusual now that, as @kit1980 mentions, it absolutely needs a comment. In the future we may want to extend the test framework to automatically generate "golden value" tests.

**Review by lezcano — COMMENTED** [review]
Left a couple comments on the testing, but otherwise it looks good.

For the complex part, you can see how it's done for cuda/UnaryLogKernels.cu. You may want to do this in a follow-up PR.

**Review by jgong5 — APPROVED** [review]
LGTM

File: test/test_unary_ufuncs.py [hunk 2]
```diff
@@ -1094,6 +1094,48 @@ def test_mish(self, device, dtype):
             rtol=rtol,
         )
 
+    @dtypes(torch.complex64, torch.complex128)
+    @onlyCPU
+    def test_log1p_complex(self, device, dtype):
```
[the remark this answers is not in the document]
Comment (lezcano) [reply test/test_unary_ufuncs.py:1099]: Actually, this may something that may be of interest in general for unary and binary ufuncs. See `test_unary_ufuncs.py`. Perhaps someone from intel may be interested in implementing this? @jgong5 The reference for the test can be either numpy or a non-contiguous tensor.

[the remark this answers is not in the document]
Comment (jgong5) [reply test/test_unary_ufuncs.py:1099]: > Actually, this may something that may be of interest in general for unary and binary ufuncs. See `test_unary_ufuncs.py`. Perhaps someone from intel may be interested in implementing this? @jgong5 The reference for the test can be either numpy or a non-contiguous tensor.

You are talking about a way of test the vector and non-vector implementation in general, right? I noticed that a lot of tests in `test_unary_ufuncs.py` already use a big size and should hit the vector path already (even though the scalar path might not be covered then). Some tests use smaller size and the vector path is not covered though. This would need a more thorough review to know the current status. We can probably make sure all the implementations can trigger the vector path and then control the `ATEN_CPU_CAPABILITY` env var to cover both vector and non-vector paths with the same test code. cc @sanchitintel and @mingfeima for their opinions.

[the remark this answers is not in the document]
Comment (lezcano) [reply test/test_unary_ufuncs.py:1099]: So, what I've seen is that there are some one-offs tests for some ops, but we could potentially have a dedicated test that tests the vectorised path against the non-vectorised explicitly. The plan to test depending on the CPU capabilities sounds great as well!

**Comment by mfkasim1** [comment]
What’s the next step for this?

**Review by lezcano — APPROVED** [review]
Great, thank you!

**Comment by mfkasim1** [comment]
Thanks @lezcano, @jgong5, @kit1980, @mruberry for your help!
